### DISCUSSION TITLE
problem with socket.accept() giving client but no data to recv

### DISCUSSION THREAD

**Discussion by gwynge** [micropython/micropython#9582] [General]
I'm new to micropython and Pico but I've been coding professionally for 35 years so not wet behind the ears.

I've got a Pico W and using latest firmware from micropython site (rp2-pico-w-20221011-unstable-v1.19.1-532-g8e912a501)

When I run the attached standard sample code and you start the web server up it starts listening on a port. When a client connects the socket.accept returns an inbound socket. the socket.recv(1024) then pulls the data in. It sends something back to the client and then closed the socket.
Immediately another inbound connection is established (although this has not been requested by the client), a new socket is returned BUT there is no data and so the socket.recv(1024) just hangs. Eventually it times out and the request data is empty. 

If another request is sent by the browser (by manually refreshing) before timeout then it appears to work correctly, but it isn't. It becomes especially obvious if you have two separate clients (different IP addresses) that are requesting data. One or other just hangs whilst the .recv continues to hang.

Can't believe something so simple is so broken! What am I doing wrong and what do I have to do to get round this?

[main.txt](https://github.com/micropython/micropython/files/9759911/main.txt)

**Comment by GM-Script-Writer-62850** [comment]
i also noticed a second client request that does not show up in browser network activity

in the async server example there is a inf loop as a result of that request
https://github.com/raspberrypi/pico-feedback/issues/279

i do not see why this script is hanging on the empty request

the empty request caused me no issues when testing that example you are going by

you could try
```
request = cl.recv(1024)
if not request:
   print('leave me alone you want NOTHING')
   cl.close()
   continue
```

running this firmware: rp2-pico-w-20220923-unstable-v1.19.1-443-g30e50ab19

**Reply by GM-Script-Writer-62850** [reply]
if you plan on severing web content everyone seems to recommend the microdot library

you will probaly want to have stuff running while the server waits on user input, so using uasyncio is a great idea unless you want to let core 2 run your web interface, if you want to use core 2 you want to use _thread

a simple use of _thread is this
```
from _thread import start_new_thread
from time import sleep
def myTask:
   while True:
      #code to blink a LED here
      sleep(1)
start_new_thread(myTask,())
while True:
   #code to blink another LED here
   sleep(1)
```

be sure to keep code running via _thread simple, it is a but buggy and does not handle everything like it should for example interrupts do not work

**Comment by gwynge** [comment]
I've discovered that it the behaviour is due to Chrome - Chrome is opening a second TCP connection, just in case it needs to use it, but not sending any data. It is described here [https://stackoverflow.com/questions/47336535/why-does-chrome-open-a-connection-but-not-send-anything](https://stackoverflow.com/questions/47336535/why-does-chrome-open-a-connection-but-not-send-anything)

When using Firefox (possibly an older version) this issue does not occur. Unfortunately there is no obvious way to detect such an "empty" connection and perhaps close it. RECV in other implementations has a Peek option to see if there is any data, Micropython doesn't seem to support this. However, I suspect that it ought to

Out of interest how does one make Microdot available on a Pico? I've found the github repository but how does one normally add something like this? (yes, I'm a proper newbie to this tech!)

**Reply by GM-Script-Writer-62850** [reply]
it would only cause a issue if you are in a infinite loop, my code i made like that ignored them perfectly, using firefox here

your code does nothing as far as processing the inbound data, that code is oh a connection here is your data have a nice day, wait for next connection

request set on line 53 should read `b''` for the blank connection

**Reply by gwynge** [reply]
The problem is that the accept() returns with a socket for inbound data but there is no data and so recv(1024) just hangs until perhaps another request in the future or a timeout. The request set on line 53 does not return anything until something (possibly Chrome) times the connection out

**Reply by GM-Script-Writer-62850** [reply]
does the same thing happen with firefox? the empty connection never caused me a issue

in my code i was doing this and had no issues, never tested chrome though, moved on to a async server setup
```
			r = cl.recv(1024)
			request = str(r).split("\\r\\n")[0].split(' ')

			if len(request) > 0:
				if request[0]=="b'GET":
					pass# stuff to do
				elif request[0]=="b'POST":
					pass# stuff to do
				else:
					print('Say what?:',request)
			else:
				print('Say what?:',request)
			cl.close()	
			print('Closed connection\n')
```

**Reply by gwynge** [reply]
Firefox doesn't seem to create the connection and send no data. Your code, in my example, will fail because the recv(1024) will never return. I've also narrowed the Chrome behaviour more - after I returned a 404 to a favicon.ico request Chrome created the empty second request every time! Clearing the cache again that second request is for favicon each time. However there is still an occasional empty request. 

I'm pretty certain it's the same thing that you're seeing in [#279](https://github.com/raspberrypi/pico-feedback/issues/279) but when you're running sync you're having to wait for recv to return, which it doesn't do. So in your doc issue you can work around the behaviour, with my issue here you can't!

**Reply by GM-Script-Writer-62850** [reply]
Firefox ( 105.0b9 (64-bit); on kunbuntu 20.04) does send a blank connection, just not 100% of the time, usually if you are looking for it you are less likely to see it and 2 min after you give up there it is
as you can see i managed to get one somehow:
[scratchpad.zip](https://github.com/micropython/micropython/files/9777369/scratchpad.zip)

```
>>> %Run -c $EDITOR_CONTENT
waiting for connection...
waiting for connection...
waiting for connection...
waiting for connection...
waiting for connection...
connected
ip = 10.0.0.171
listening on ('0.0.0.0', 80)


Waiting for connection
Client connected from ('10.0.0.50', 34748)
request= b''
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 55896)
request= b'GET / HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nDNT: 1\r\nConnection: keep-alive\r\nUpgrade-Insecure-Requests: 1\r\n\r\n'
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 55906)
request= b'GET /favicon.ico HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: image/avif,image/webp,*/*\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nReferer: http://10.0.0.171/\r\nDNT: 1\r\nConnection: keep-alive\r\n\r\n'
sending html
connection closed normally


Waiting for connection
```
and right after i give up another one:
```
>>> %Run -c $EDITOR_CONTENT
waiting for connection...
waiting for connection...
waiting for connection...
waiting for connection...
waiting for connection...
connected
ip = 10.0.0.171
listening on ('0.0.0.0', 80)


Waiting for connection
Client connected from ('10.0.0.50', 33188)
request= b'GET / HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nDNT: 1\r\nConnection: keep-alive\r\nUpgrade-Insecure-Requests: 1\r\n\r\n'
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 42382)
request= b'GET / HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nDNT: 1\r\nConnection: keep-alive\r\nUpgrade-Insecure-Requests: 1\r\n\r\n'
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 42388)
request= b'GET / HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nDNT: 1\r\nConnection: keep-alive\r\nUpgrade-Insecure-Requests: 1\r\n\r\n'
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 42396)
request= b'GET / HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nDNT: 1\r\nConnection: keep-alive\r\nUpgrade-Insecure-Requests: 1\r\n\r\n'
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 38876)
request= b'GET / HTTP/1.1\r\nHost: 10.0.0.171\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:105.0) Gecko/20100101 Firefox/105.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\nAccept-Language: en-US,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nDNT: 1\r\nConnection: keep-alive\r\nUpgrade-Insecure-Requests: 1\r\n\r\n'
sending html
connection closed normally


Waiting for connection
Client connected from ('10.0.0.50', 38860)
request= b''
sending html
connection closed normally


Waiting for connection
```
